// vuepress-theme-plume/lib/client/index.js
// src/client/index.ts
export * from "../shared/index.js";
import { default as default2 } from "./components/global/VPBadge.vue";
import { default as default3 } from "./components/global/VPCard.vue";
import { default as default4 } from "./components/global/VPCardGrid.vue";
import { default as default5 } from "./components/Home/VPHomeBanner.vue";
import { default as default6 } from "./components/Home/VPHomeBox.vue";
import { default as default7 } from "./components/Home/VPHomeCustom.vue";
import { default as default8 } from "./components/Home/VPHomeFeatures.vue";
import { default as default9 } from "./components/Home/VPHomeHero.vue";
import { default as default10 } from "./components/Home/VPHomeProfile.vue";
import { default as default11 } from "./components/Home/VPHomeTextImage.vue";
import { default as default12 } from "./components/VPButton.vue";
import { default as default13 } from "./components/VPImage.vue";
import { default as default14 } from "./components/VPLink.vue";
export * from "./composables/index.js";
import { default as default15 } from "./config.js";
import { default as default16 } from "./layouts/Layout.vue";
import { default as default17 } from "./layouts/NotFound.vue";
export * from "./utils/index.js";
export {
  default16 as Layout,
  default17 as NotFound,
  default2 as VPBadge,
  default12 as VPButton,
  default3 as VPCard,
  default4 as VPCardGrid,
  default5 as VPHomeBanner,
  default6 as VPHomeBox,
  default7 as VPHomeCustom,
  default8 as VPHomeFeatures,
  default9 as VPHomeHero,
  default10 as VPHomeProfile,
  default11 as VPHomeTextImage,
  default13 as VPImage,
  default14 as VPLink,
  default15 as plumeClientConfig
};
