// @pengzhanbo/utils/dist/cjs/object.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.deepFreeze = deepFreeze;
exports.deepMerge = deepMerge;
exports.deepMergeWithArray = deepMergeWithArray;
exports.hasOwn = hasOwn;
exports.isKeyof = isKeyof;
exports.objectGet = objectGet;
var _is = require("./is.cjs");
function hasOwn(obj, key) {
  return obj === null ? false : Object.prototype.hasOwnProperty.call(obj, key);
}
function deepFreeze(obj) {
  if ((0, _is.isArray)(obj)) {
    obj.forEach(item => deepFreeze(item));
  } else if ((0, _is.isObject)(obj)) {
    Object.freeze(obj);
    Object.keys(obj).forEach(key => deepFreeze(obj[key]));
  }
  return obj;
}
function isKeyof(obj, key) {
  return key in obj;
}
function objectGet(source, path) {
  const keys = path.replace(/\[['"]?(.+?)['"]?\]/g, ".$1").split(".");
  let res = source;
  for (const k of keys) res = res?.[k];
  return res;
}
function deepMerge(target, ...sources) {
  if (!sources.length) return target;
  const source = sources.shift();
  if (source === void 0) return target;
  if (isMergableObject(target) && isMergableObject(source)) {
    Object.keys(source).forEach(key => {
      if (key === "__proto__" || key === "constructor" || key === "prototype") return;
      if (isMergableObject(source[key])) {
        if (!target[key]) target[key] = {};
        deepMerge(target[key], source[key]);
      } else {
        target[key] = source[key];
      }
    });
  }
  return deepMerge(target, ...sources);
}
function deepMergeWithArray(target, ...sources) {
  if (!sources.length) return target;
  const source = sources.shift();
  if (source === void 0) return target;
  if (Array.isArray(target) && Array.isArray(source)) target.push(...source);
  if (isMergableObject(target) && isMergableObject(source)) {
    Object.keys(source).forEach(key => {
      if (key === "__proto__" || key === "constructor" || key === "prototype") return;
      if (Array.isArray(source[key])) {
        if (!target[key]) target[key] = [];
        deepMergeWithArray(target[key], source[key]);
      } else if (isMergableObject(source[key])) {
        if (!target[key]) target[key] = {};
        deepMergeWithArray(target[key], source[key]);
      } else {
        target[key] = source[key];
      }
    });
  }
  return deepMergeWithArray(target, ...sources);
}
function isMergableObject(item) {
  return (0, _is.isObject)(item) && !Array.isArray(item);
}
